storage/src/checker.rs

```diff
@@ -208,4 +271,68 @@ mod test {
         let complement = visited.complement();
         assert_eq!(complement.into_iter().collect::<Vec<_>>(), vec![]);
     }
+
+    #[test]
+    // This test creates a linear set of stored areas and free some of them.
+    // When traversing it should break consecutive areas.
+    fn find_leaked_areas() {
+        use rand::Rng;
+        use rand::seq::IteratorRandom;
+
+        let mut rng = crate::test_utils::seeded_rng();
+
+        let memstore = MemStore::new(vec![]);
+        let mut nodestore = NodeStore::new_empty_committed(memstore.into()).unwrap();
+
+        let num_areas = 10;
+
+        // randomly insert areas into the nodestore
+        let mut high_watermark = NodeStoreHeader::SIZE;
+        let mut stored_areas = Vec::new();
+        for _ in 0..num_areas {
+            let (area_size_index, area_size) =
+                AREA_SIZES.iter().enumerate().choose(&mut rng).unwrap();
+            let area_addr = LinearAddress::new(high_watermark).unwrap();
+            test_write_random_stored_area(&nodestore, area_size_index as AreaIndex, high_watermark);
+            stored_areas.push((area_addr, area_size_index as AreaIndex, *area_size));
+            high_watermark += *area_size;
+        }
+        test_write_header(&mut nodestore, high_watermark, None, FreeLists::default());
+
+        // randomly pick some areas as leaked
+        let mut leaked = Vec::new();
+        let mut expected_free_areas = HashMap::new();
+        let mut area_to_free = None;
+        for (area_start, area_size_index, area_size) in &stored_areas {
+            if rng.random::<f64>() < 0.5 {
+                // we free this area
+                expected_free_areas.insert(*area_start, *area_size_index);
+                let area_to_free_start = area_to_free.map_or(*area_start, |(start, _)| start);
+                let area_to_free_end = area_start.checked_add(*area_size).unwrap();
+                area_to_free = Some((area_to_free_start, area_to_free_end));
+            } else {
+                // we are not freeing this area, free the aggregated areas before this one
+                if let Some((start, end)) = area_to_free {
+                    leaked.push(start..end);
+                    area_to_free = None;
+                }
+            }
+        }
+        if let Some((start, end)) = area_to_free {
+            leaked.push(start..end);
+        }
+
+        // check the leaked areas
+        let leaked_areas = leaked
+            .into_iter()
+            .map(|range| nodestore.find_leaked_areas(range))
+            .collect::<Result<Vec<_>, CheckerError>>()
+            .unwrap()
+            .into_iter()
+            .flatten()
+            .collect::<HashMap<_, _>>();
```

Comment:
This looks really familiar. Maybe break this iterator out into a helper?

I realize this is in a test but you can call the non-test code from there.